### PR TITLE
Align implementation of InterPodAffinity, TaintToleration, PodTopologySpread plugins with other plugins in nodeorder and add unit tests

### PR DESCRIPTION
**Pull request by JesseStutler**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Add more unit tests in this version after we merged: 
https://github.com/volcano-sh/volcano/pull/4804
https://github.com/volcano-sh/volcano/pull/4848
Besides, current implementation of InterPodAffinity, TaintToleration, PodTopologySpread and doesn't align with other plugins, they should be added to scorePlugins map when they're successfully initialized, and they should be initialized only weight is not 0

#### Which issue(s) this PR fixes:

Related https://github.com/volcano-sh/volcano/issues/4957

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

File: pkg/scheduler/util/predicate_helper_test.go [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+				Job:       "job1",
+				TaskRole:  "worker",
+				Namespace: "ns",
+				Name:      "task",
+			},
+			nodes: []*api.NodeInfo{
+				{Name: "node2"},
+			},
+			nodesInShard:     sets.New[string]("node1"),
+			shardingMode:     commonutil.HardShardingMode,
```
Comment (qi-min) [inline pkg/scheduler/util/predicate_helper_test.go:118]: maybe a test case for softmode can be added to make sure nodes will not be exluded in soft mode,
and another test case for hardmode to verify nodes in shard will not be excluded

File: pkg/scheduler/plugins/nodeorder/nodeorder_test.go [hunk 2]
```diff
@@ -139,3 +348,88 @@ func TestNodeOrderPlugin(t *testing.T) {
...
+			name:                  "all weights zero",
+			weight:                priorityWeight{},
+			expectNodeOrderKeys:   []string{},
+			expectNoNodeOrderKeys: []string{noderesources.Name + "_LeastAllocated", noderesources.Name + "_MostAllocated", noderesources.BalancedAllocationName, nodeaffinity.Name, imagelocality.Name},
+			expectScorePlugins: []string{
+				interpodaffinity.Name,
+				tainttoleration.Name,
+				podtopologyspread.Name,
+			},
+			expectNotInScorePlugins: []string{},
```
[the remark this answers is not in the document]
Comment (JesseStutler) [reply pkg/scheduler/plugins/nodeorder/nodeorder_test.go]: You're right, solved

File: pkg/scheduler/plugins/nodeorder/nodeorder_test.go [hunk 3]
```diff
@@ -101,6 +118,192 @@ func TestNodeOrderPlugin(t *testing.T) {
...
+		{
+			TestCommonStruct: uthelper.TestCommonStruct{
+				Name: "pod topology spread prefers node to reduce skew",
+				PodGroups: []*schedulingv1.PodGroup{
+					util.BuildPodGroup("pg1", "c1", "c1", 0, nil, schedulingv1.PodGroupInqueue),
+				},
+				Pods: []*v1.Pod{
+					// Existing pod on n1 (zone1)
+					util.BuildPod("c1", "p1", "n1", v1.PodRunning, api.BuildResourceList("1", "1G"), "pg1", map[string]string{"app": "foo"}, make(map[string]string)),
+					// Existing pod on n2 (zone1)
```
[the remark this answers is not in the document]
Comment (JesseStutler) [reply pkg/scheduler/plugins/nodeorder/nodeorder_test.go]: You're right, solved

File: pkg/scheduler/plugins/nodeorder/nodeorder.go [hunk 4]
```diff
@@ -322,23 +340,43 @@ func (pp *NodeOrderPlugin) BatchNodeOrderFn(task *api.TaskInfo, nodeInfo []*api.
 	}
 	nodeScores := make(map[string]float64, len(nodes))
 
-	podAffinityScores, podErr := interPodAffinityScore(pp.ScorePlugins[interpodaffinity.Name].(*interpodaffinity.InterPodAffinity), state, task.Pod, nodeInfos, pp.weight.podAffinityWeight)
-	if podErr != nil {
-		return nil, podErr
+	var podAffinityScores map[string]float64
+	var err error
+	if pp.ScorePlugins[interpodaffinity.Name] != nil {
```
Comment (hajnalmt) [inline pkg/scheduler/plugins/nodeorder/nodeorder.go:345]: These if branches doesn't look proper, you are dereferencing the `Name`, but you are not initializing the plugin if the weight is 0 so this will cause a panic if the weight is 0 right?

File: pkg/scheduler/plugins/nodeorder/nodeorder_test.go [hunk 5]
```diff
@@ -101,6 +118,192 @@ func TestNodeOrderPlugin(t *testing.T) {
...
+				Queues: []*schedulingv1.Queue{
+					util.BuildQueue("c1", 1, nil),
+				},
+				ExpectBindsNum: 1,
+				ExpectBindMap: map[string]string{
+					"c1/p4": "n3", // n3 (zone3) is empty, n2 (zone2) has 1 pod and n1(zone1) has 2 pods. spread is better on n3.
+				},
+			},
+			PodTopologySpreadWeight: 10,
+		},
```
Comment (hajnalmt) [inline pkg/scheduler/plugins/nodeorder/nodeorder_test.go:306]: Maybe adding a runtime test with zero weights and all the pods is needed here (the "all weights zero" test checks only the map keys, doesn't call the function directly). It should bring up my theory that this will panic out: 
```
{
    TestCommonStruct: uthelper.TestCommonStruct{
        Name: "zero weight for batch scoring plugins with constraint pods",
        PodGroups: []*schedulingv1.PodGroup{
            util.BuildPodGroup("pg1", "c1", "c1", 0, nil, schedulingv1.PodGroupInqueue),
        },
        Pods: []*v1.Pod{
            util.BuildPodWithAffinity("c1", "p1", "", v1.PodPending,
                api.BuildResourceList("1", "1G"), "pg1",
                make(map[string]string), make(map[string]string),
                &v1.Affinity{
                    PodAffinity: &v1.PodAffinity{
                        PreferredDuringSchedulingIgnoredDuringExecution: []v1.WeightedPodAffinityTerm{
                            {
                                Weight: 100,
                                PodAffinityTerm: v1.PodAffinityTerm{
                                    LabelSelector: &metav1.LabelSelector{
                                        MatchLabels: map[string]string{"app": "test"},
                                    },
                                    TopologyKey: "kubernetes.io/hostname",
                                },
                            },
                        },
                    },
                }),
            util.BuildPodWithTolerations("c1", "p2", "", v1.PodPending,
                api.BuildResourceList("1", "1G"), "pg1",
                make(map[string]string), make(map[string]string),
                []v1.Toleration{
                    {
                        Key:      "test-key",
                        Operator: v1.TolerationOpEqual,
                        Value:    "test-value",
                        Effect:   v1.TaintEffectNoSchedule,
                    },
                }),
            util.BuildPodWithTopologySpreadConstraints("c1", "p3", "", v1.PodPending,
                api.BuildResourceList("1", "1G"), "pg1",
                make(map[string]string), make(map[string]string),
                []v1.TopologySpreadConstraint{
                    {
                        MaxSkew:           1,
                        TopologyKey:       "zone",
                        WhenUnsatisfiable: v1.ScheduleAnyway,
                        LabelSelector: &metav1.LabelSelector{
                            MatchLabels: map[string]string{"app": "test"},
                        },
                    },
                }),
        },
        Nodes: []*v1.Node{
            util.BuildNode("n1", api.BuildResourceList("4", "8Gi",
                []api.ScalarResource{{Name: "pods", Value: "10"}}...),
                make(map[string]string)),
            util.BuildNode("n2", api.BuildResourceList("8", "16Gi",
                []api.ScalarResource{{Name: "pods", Value: "10"}}...),
                make(map[string]string)),
        },
        Queues: []*schedulingv1.Queue{
            util.BuildQueue("c1", 1, nil),
        },
        ExpectBindsNum: 3,
        // All pods should schedule successfully based on LeastRequestedWeight
        ExpectBindMap: map[string]string{
            "c1/p1": "n2",
            "c1/p2": "n2",
            "c1/p3": "n2",
        },
    },
    LeastRequestedWeight:    1,
    MostRequestedWeight:     0,
    PodAffinityWeight:       0,
    TaintTolerationWeight:   0,
    PodTopologySpreadWeight: 0,
},
```

File: pkg/scheduler/plugins/nodeorder/nodeorder.go (hunk 4 above)
Comment (JesseStutler) [reply pkg/scheduler/plugins/nodeorder/nodeorder.go:345]: You mean `pp.ScorePlugins` will be nil? But currently we will always the initialize ScorePlugins maps during `InitPlugins`, so we will not cause panic here. And if the weight is 0, so `pp.ScorePlugins[interpodaffinity.Name]` will return a nil, and the `if` protects the situation

Comment (hajnalmt) [reply pkg/scheduler/plugins/nodeorder/nodeorder.go:345]: No, `interpodAffinity` is `nil` when weight is 0, so `interpodAffinity.Name ` will cause a panic in that case I think.

Comment (JesseStutler) [reply pkg/scheduler/plugins/nodeorder/nodeorder.go:345]: The `interpodaffinity` here is a package name, instead of a variable name, so it will not cause panic I think 😄

Comment (hajnalmt) [reply pkg/scheduler/plugins/nodeorder/nodeorder.go:345]: Oh you are right sorry,  the cases tricked me thank you!

**Comment by JesseStutler** [comment]
Oh I'm still working on add @qi-min 's suggested tests but didn't notice it was already merged: https://github.com/volcano-sh/volcano/pull/4983#discussion_r2720378975 😳, I will submit a new PR to fix it

File: pkg/scheduler/util/predicate_helper_test.go (hunk 1 above)
Comment (JesseStutler) [reply pkg/scheduler/util/predicate_helper_test.go:118]: This pr was already merged, I will raise a new one
